### PR TITLE
Add CurieBle class for easy use, change implementation to containing class

### PR DESCRIPTION
**Pull request by mwmckee**
Implementation of BleSerial functions was moved to a separate class, which DfRobotBleSerial and the new CurieBleSerial draw from. This consolidates the implementation for BleSerial into one class, and allows for specific devices to simply wrap the main implementation to provide easy use constructors for specific BLE GUIDs. 

This implementation style was necessary as C++/CX does not allow for custom base classes, so the CurieBleSerial and DfRobotBleSerial cannot simply inherit BleSerial as a base class and override only what is necessary.

Right now, there isn't a clear path forward with making any sort of polymorphic constructor for BleSerial, which would allow us to avoid creating a full wrapper class for each new BLE GUID that we want to provide simple support for. This implementation is the next best thing given that.

Further, this implementation allows for users to input their own GUIDs by creating a BleSerial class directly, so that our library won't need changed to allow new ones to be used.

### CODE REVIEW

**Comment by mwmckee** [comment]
@zafields @JimGaleForce @sidwarkd can you all give this a review?

File: source/BleSerial.h [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    ///Begins an asyncronous request for all Bluetooth LE devices that are paired and may be used to attempt a device connection.
+    ///</summary>
+    static
+    Windows::Foundation::IAsyncOperation<Windows::Devices::Enumeration::DeviceInformationCollection ^> ^
+    listAvailableDevicesAsync(
+        void
+        );
+
+private:
+    Platform::Guid BLE_SERVICE_UUID;
```
Comment (zfields) [inline source/BleSerial.h]: make the GUIDs const if you can

File: source/CurieBleSerial.cpp [hunk 2]
```diff
@@ -0,0 +1,378 @@
+/*
```
Comment (zfields) [inline source/CurieBleSerial.cpp]: Do we need this file, or can we just use `inline` functions in the `.h` file?

File: source/CurieBleSerial.h [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+    virtual
+        ~CurieBleSerial(
+            void
+        )
+    {
+        _bleSerial->~BleSerial();
+    };
+
+    virtual
```
Comment (zfields) [inline source/CurieBleSerial.h]: mark all the pure call through functions as `virtual inline`

File: source/DfRobotBleSerial.cpp [hunk 4]
```diff
@@ -105,7 +105,7 @@ DfRobotBleSerial::available(
     else {
```
Comment (zfields) [inline source/DfRobotBleSerial.cpp]: This file should go the way of `CurieBleSerial.cpp`...

File: source/CurieBleSerial.cpp (hunk 2 above)
Comment (mwmckee) [inline source/CurieBleSerial.cpp]: We should be able to inline. I removed this file and the DfRobotBleSerial file from the project, but forgot to yank them from this version in the source control. Will fix with next iteration.

File: source/DfRobotBleSerial.h [hunk 5]
```diff
@@ -1,236 +1,279 @@
...
+                {
+                    _bleSerial = ref new BleSerial(
+                        device_,
+                        Platform::Guid{ 0x6E400001, 0xB5A3, 0xF393, (0xE0, 0xA9, 0xE5, 0x0E, 0x24, 0xDC, 0xCA, 0x9E) },
+                        Platform::Guid{ 0x6E400002, 0xB5A3, 0xF393, (0xE0, 0xA9, 0xE5, 0x0E, 0x24, 0xDC, 0xCA, 0x9E) },
+                        Platform::Guid{ 0x6E400003, 0xB5A3, 0xF393, (0xE0, 0xA9, 0xE5, 0x0E, 0x24, 0xDC, 0xCA, 0x9E) }
+                    );
+                };
+
+                virtual
```
Comment (zfields) [inline source/DfRobotBleSerial.h]: Be sure to `inline` all call through functions.

File: source/CurieBleSerial.h [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    Windows::Foundation::IAsyncOperation<Windows::Devices::Enumeration::DeviceInformationCollection ^> ^
+    listAvailableDevicesAsync(
+        void
+        )
+    {
+        return BleSerial::listAvailableDevicesAsync();
+    };
+
+private:
+    BleSerial^ _bleSerial;
```
Comment (zfields) [inline source/CurieBleSerial.h]: Make this `const`

File: source/CurieBleSerial.h [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+    virtual event IStreamConnectionCallbackWithMessage ^ConnectionLost;
+    virtual event IStreamConnectionCallbackWithMessage ^ConnectionFailed;
+
+    ///<summary>
+    ///A constructor which accepts a string corresponding to a device name or ID to connect to.
+    ///</summary>
+    CurieBleSerial(
+        Platform::String ^device_name_
+    )
+    {
```
Comment (zfields) [inline source/CurieBleSerial.h]: instantiate `_bleSerial` after the `:` and before the `{`

File: source/CurieBleSerial.h (hunk 6 above)
Comment (mwmckee) [inline source/CurieBleSerial.h]: Unfortunately, it looks like this sort of object nesting prohibits creating these as const objects.

File: source/CurieBleSerial.cpp (hunk 2 above)
Comment (mwmckee) [inline source/CurieBleSerial.cpp]: Following up - removing these files totally prevents the proper metadata from being generated, which in turns prevents CurieBleSerial and DfRobotBleSerial from being accessed. I've added .cpp files with the minimal needed code to force correct compilation.

File: source/CurieBleSerial.h (hunk 6 above)
Comment (zfields) [inline source/CurieBleSerial.h]: It shouldn't as long as you instatiate it in the constructor

File: source/BleSerial.h [hunk 8]
```diff
@@ -0,0 +1,244 @@
+/*
+    Copyright(c) Microsoft Open Technologies, Inc. All rights reserved.
```
Comment (JimGaleForce) [inline source/BleSerial.h]: still need these copyrights?

**Comment by JimGaleForce** [comment]
looks good (with replying to zak's comments)

File: source/BleSerial.h (hunk 8 above)
Comment (mwmckee) [inline source/BleSerial.h]: Probably not, but I'd rather keep them in for now, so we don't need to run it by LCA

File: source/CurieBleSerial.h [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+    void
+    begin(
+        void
+        )
+    {
+        //baud rate and serial configuration are unnecessary for BLE connections as we are not using asyncronous TTL
+        begin(NULL, SerialConfig::NONE);
+    }
+
+    virtual
```
Comment (zfields) [inline source/CurieBleSerial.h]: missing `inline`

File: source/CurieBleSerial.cpp [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+    THE SOFTWARE.
+*/
+
+/*
+    Implementation of DfRobotBleSerial is contained in the DfRobotBleSerial.h
```
Comment (zfields) [inline source/CurieBleSerial.cpp]: Comment should say "CurieBleSerial" instead of "DfRobotBleSerial"

File: source/DfRobotBleSerial.cpp [hunk 11]
```diff
@@ -22,455 +22,15 @@
...
-void
-DfRobotBleSerial::rxCallback(
-    GattCharacteristic ^sender,
-    GattValueChangedEventArgs ^args
-    )
-{
-    // Extract data into workable form from parameters
-    Platform::Array<byte> ^rx_data = ref new Platform::Array<byte>(args->CharacteristicValue->Length);
-    DataReader::FromBuffer(args->CharacteristicValue)->ReadBytes(rx_data);
+/*
```
Comment (zfields) [inline source/DfRobotBleSerial.cpp]: Great documentation!

**Pull request merged**